adding basic tests #45

lhein opened the pull request:
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

apupier commented on src/commands/CreateVDBCommand.ts:
```diff
@@ -53,17 +53,17 @@ export function createVDBCommand() {
 		if (fileName && fileName.length>0) {
 			try {
-				let templatePath = path.join(extension.pluginResourcesPath, "vdb_template.yaml");
+				let templatePath = templateFolder ? path.join(templateFolder, "vdb_template.yaml") : path.join(extension.pluginResourcesPath, "vdb_template.yaml");
```
I think it would be best to remove the logic from here and force passing the templateFolder for all callers.

Also I don't understand why we have 2 different cases. We want to tets with different templates in test? Why?

apupier commented on src/test/command.create.vdb.test.ts:
```diff
@@ -0,0 +1,50 @@
+'use strict';
```
missing header

apupier commented on src/test/utils.test.ts:
```diff
@@ -0,0 +1,110 @@
+'use strict';
```
missing header

apupier commented on src/test/utils.test.ts:
```diff
@@ -0,0 +1,110 @@
+		expect(utils.validateName('xyz')).to.not.be.undefined;
+		
+		// positive test
+		expect(utils.validateName('test')).to.be.undefined;		
```
why adding this one inside the same test? you can add it in a single different test for testing "valid naming doesn't report validation error"

apupier commented on src/test/command.create.vdb.test.ts:
```diff
@@ -0,0 +1,50 @@
+import * as extension from '../extension';
+import { handleVDBCreation } from '../commands/CreateVDBCommand';
+
+const YAML = require('yaml');
```
this is unused

apupier commented on src/test/utils.test.ts:
```diff
@@ -0,0 +1,110 @@
+
+describe('Utils Functional Tests', function () {
+
+	let sandbox: sinon.SinonSandbox;
```
the sinon sandbox can be removbed as no spy/stub are usd

apupier commented on src/test/utils.test.ts:
```diff
@@ -0,0 +1,110 @@
+		expect(utils.validateName('myTest$1')).to.not.be.undefined;
+		expect(utils.validateName('test#1')).to.not.be.undefined;
+		expect(utils.validateName('my.Test')).to.not.be.undefined;
+		expect(utils.validateName('test/1')).to.not.be.undefined;
```
better to split test to give more precise naming

apupier commented on src/test/utils.test.ts:
```diff
@@ -0,0 +1,110 @@
+		expect(yamlDoc.metadata.name).to.be.equal(yamlDoc2.metadata.name);
+		expect(yamlDoc.spec.env.length).to.be.equal(yamlDoc2.spec.env.length);
+		expect(yamlDoc2.spec.build.source.ddl).to.not.be.undefined;
+		fs.unlinkSync(fpTest);
```
what is the purpose of this unlinkSync?

apupier commented on src/test/utils.test.ts:
```diff
@@ -0,0 +1,110 @@
+		let fpOrig: string = path.resolve(__dirname, '../../testFixture', `${name}.yaml`);
+		let fpTest: string = path.resolve(__dirname, '../../testFixture', `${name}2.yaml`);
+		let yamlDoc:IDVConfig = utils.loadModelFromFile(fpOrig);
+		expect(yamlDoc).to.not.be.undefined;
```
adding message will help to debug and spot the specific place the test is failing

apupier commented on .editorconfig:
```diff
@@ -0,0 +1,14 @@
+# Normalizes editor configuration.
```
this file is not related to this PR. Would be best in another PR

lhein replied:
I will keep it in this PR.

apupier commented on .vscode/settings.json:
```diff
@@ -2,5 +2,6 @@
 	"prettier.tabWidth": 4,
-	"prettier.useTabs": true
+	"prettier.useTabs": true,
+	"typescript.tsdk": "node_modules/typescript/lib",
```
what is the purpose of this configuration?

lhein replied:
This is injected when you manually configure the TypeScript version through the UI to use your local typescript version. This is part of the configuration steps for the TSLinter if I remember correctly.

apupier commented on gulpfile.js:
```diff
@@ -4,7 +4,7 @@ var gulp = require("gulp");
 
 gulp.task("tslint", () => {
-    return gulp.src(['**/*.ts', '!**/*.d.ts', '!node_modules/**'])
+    return gulp.src(['src/**/*.ts', '!**/*.d.ts', '!node_modules/**'])
```
Why is it related to the additional tests?

lhein replied:
it is not directly related but I keep it here.

apupier replied:
what is the goal of the modification then?

apupier replied:
ah finally removed, my bad

apupier commented on package.json:
```diff
@@ -161,6 +161,7 @@
 		"sinon-chai": "^3.5.0",
 		"tslint": "^6.1.0",
 		"typescript": "^3.8.3",
+		"typescript-tslint-plugin": "^0.5.5",
```
would be better to add in a different PR, it doesn't sound linked to "adding basic tests"

lhein replied:
I added the tslint plugin to help fix obvious problems in code and formatting. Keeping it here.

apupier commented on .vscode/launch.json:
```diff
@@ -21,9 +21,11 @@
+				"${workspaceRoot}/testFixture/"],
 			"stopOnEntry": false,
 			"sourceMaps": true,
+			"preLaunchTask": "npm: watch",
```
can you explain the purpose of adding the watch here?

lhein replied:
actually I have no clue why that has been changed. Guess I will remove it.

apupier commented on src/test/outputchannel.log.test.ts:
```diff
@@ -14,36 +14,35 @@
+			windowSpy
+				.should.have.been.calledOnce
+				.and.been.calledWith('DataVirt Extension');
+		});
```
we need to ensure that it has not been called before calling the log method. In case, there is already a log done in another way.

apupier replied:
reported to new code https://github.com/jboss-fuse/vscode-datavirt/pull/84/files#r397300938

apupier commented on src/test/outputchannel.log.test.ts:
```diff
@@ -14,36 +14,35 @@
 
-describe("Test Extension OutputChannel", function() {
+describe('Extension OutputChannel', () => {
+	context('OutputChannelLogger', () => {
```
as there is a single test in this file. Do we really need a "context"?

apupier commented on src/test/utils.test.ts:
```diff
@@ -0,0 +1,165 @@
+	 * disabled because broken probably due to incorrect path for testFixture folder
+	 */
+	// context('Load/Save of a VDB file', () => {
+	// 	it('should match the vdb model contents between a save and reload to/from a vdb file', () => {
```
better to use .skip instead of commenting out

apupier replied:
this one sounds like an important test. Either need to implement it in this PR or create a specific subtask for it

apupier commented on tslint.json:
```diff
@@ -1,6 +1,17 @@
+		"arrow-parens": true,
+		"eofline": true,
+		"no-console": false
+    }
```
unrelated to "adding test" PR

lhein replied:
it is not directly related but it helps improving code quality.

apupier replied:
it means that we will need to discuss the parameters in this PR too.

lhein replied:
we should discuss that at some point but I don't see how that holds this PR back. We agreed that we need to define a set of rules to use in all our projects. Whenever we did that we can raise issues and update it.

apupier commented on Contributing.md:
```diff
@@ -2,6 +2,20 @@
+This project uses Mocha and Chai for testing.
+Visit http://mochajs.org and http://chaijs.com for details.
+
+Code coverage generated by http://istanbul.js.org
```
unrelated to "adding test"

lhein replied:
keeping it here anyways because it improves the documentation

lhein commented:
I see an overlap with QE UI Testing. My idea was not to test user workflows in the unit tests
but rather to write tests for the underlying functions that are called by the user actions.
While I think its good to have a test that creates a VDB I am rather unsure if it makes sense to test if the file appears in the tree node and the ddl editor is opened afterwards. Those are more like integration tests. What do u think?

apupier commented on src/test/outputchannel.log.test.ts:
```diff
@@ -14,36 +14,33 @@
+		log('This is a test!');
+		windowSpy
+			.should.have.been.calledOnce
+			.and.been.calledWith('DataVirt Extension');
```
we need to ensure that it has not been called before calling the log method. In case, there is already a log done in another way.

(reporting again as it is marked as outdated as code has been changed since my comment

apupier commented:
> I see an overlap with QE UI Testing. My idea was not to test user workflows in the unit tests
> but rather to write tests for the underlying functions that are called by the user actions.
> While I think its good to have a test that creates a VDB I am rather unsure if it makes sense to test if the file appears in the tree node and the ddl editor is opened afterwards. Those are more like integration tests. What do u think?

agree that there is an overlap.
Fine to delegate to UI tests a part of the testing, we will need to wait for it to be merged before declaring the use case done.
it remains useful to have basic test at Mocha test level, they are automatically launched when launching test locally, so easier to spot them. They are usually faster too.

apupier commented:
current error on Travis:

```
  1) Utils
       Load/Save of a VDB file
         should match the vdb model contents between a save and reload to/from a vdb file:
     TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type undefined
  	at validateString (internal/validators.js:112:11)
  	at Object.join (path.js:1040:7)
  	at Object.validateFileNotExisting (out/utils.js:9:5376)
  	at Context.<anonymous> (out/test/utils.test.js:9:7886)
  	at processImmediate (internal/timers.js:439:21)
```

apupier commented on src/extension.ts:
```diff
@@ -141,4 +137,12 @@ function handleVisibleEditorChanges(event) {
+}
+
+export function fillDataTypes(): void {
+	DATASOURCE_TYPES.set('SpringBoot', new SpringDataSource(TEMPLATE_NAME));
```
typo SpringBoot -> Spring Boot

apupier commented on src/test/commands.datasource.entry.test.ts:
```diff
@@ -0,0 +1,743 @@
+			const eName: string = 'NEWKEY';
+			const eValue: string = 'NEWVALUE';
+			createDSEntryCommand.handleDataSourceEntryCreation(dvConfig, undefined, f, eName, eValue)
+				.then( (created) => {
+					if (created) {
+						done(new Error('Execution of the Create DataSource Entry command returned true, but should not.'));
+					} else {
+						done();
+					}
+				})
```
Sonar is mentioning that this inlined function is repeated several times in the file.  Would be better to factorize it.

lhein replied:
I am not going to change that.

apupier replied:
why not avoiding duplicate code in the same file?

lhein replied:
because it will remove readability from the test itself.

apupier replied:
I find more readable something like:

```
createDSEntryCommand.handleDataSourceEntryCreation(dvConfig, undefined, f, eName, eValue)
	.then(checkDataSourceEntryCreationReturnTrue(done))
	.catch( (err) => {
		done(err);
	});
```
than

```
createDSEntryCommand.handleDataSourceEntryCreation(dvConfig, undefined, f, eName, eValue)
	.then( (created) => {
		if (created) {
			done(new Error('Execution of the Create DataSource Entry command returned true, but should not.'));
		} else {
			done();
		}
	})
```

apupier replied:
it has the following advantages:
- we can give a clear name to the function
- it avoids duplicated code
- it limits depth level of nested code
- shorter code

apupier commented on src/test/commands.datasource.entry.test.ts:
```diff
@@ -0,0 +1,743 @@
+										} else {
+											done(new Error('Execution of the Create DataSource Entry command returned false'));
+										}
+									})
```
Sonar is mentioning that this inlined function is repeated several times in the file. Would be better to factorize it.

apupier commented on src/test/commands.datasource.entry.test.ts:
```diff
@@ -0,0 +1,743 @@
+					} else {
+						done();
+					}
+				})
```
Sonar is mentioning that this inlined function is repeated several times in the file. Would be better to factorize it.

lhein commented:
Regarding TSLInt Deprecation:
![image](https://user-images.githubusercontent.com/435853/78115634-5cde0700-7403-11ea-8f0f-579f31503786.png)
I cannot see where this is marked deprecated.

apupier commented:
> * it is marked that TSLint is deprecated https://github.com/palantir/tslint . might be best to start using ESLint directly instead of using TSLint



> I cannot see where this is marked deprecated.

by following the provided link:
![Screenshot from 2020-04-01 10-29-26](https://user-images.githubusercontent.com/1105127/78115847-b0505500-7403-11ea-9f8d-8167d38cce60.png)

lhein commented:
> to really use .editorconfig, I guess we need to have a specific extension to be installed. This extension should be added in recommended extension

This is described in the appropriate contributing doc file. Though defining it in a user friendly way is great. Thanks for the link. I will add that file.

lhein commented:
> > * it is marked that TSLint is deprecated https://github.com/palantir/tslint . might be best to start using ESLint directly instead of using TSLint
> 
> > I cannot see where this is marked deprecated.
> 
> by following the provided link:
> ![Screenshot from 2020-04-01 10-29-26](https://user-images.githubusercontent.com/1105127/78115847-b0505500-7403-11ea-9f8d-8167d38cce60.png)

I am not using that plugin. The used plugin is https://github.com/Microsoft/typescript-tslint-plugin

lhein commented:
>     * when closing the ddl editor without saving, there is no warning that changes will be ignored (really close to normal path mentioned as important to test initially)

I raised a new issue for this problem. https://github.com/jboss-fuse/vscode-datavirt/issues/102

lhein commented:
>     * mentioning issues with other features than "create vdb files" that are tested here as I would expect it to be working as automated test integrated in the same PR and so maybe handled in this PR.

Tbh I have no clue what you mean. Could you elaborate?

apupier commented:
> I am not using that plugin. The used plugin is https://github.com/Microsoft/typescript-tslint-plugin

this is the VS Code extension. The VS Code extension is using the deprecated linter. In readme of the extension, on the first line, the second link for TSLint is pointing to the deprecated linter
![deprecatedtslint](https://user-images.githubusercontent.com/1105127/78116783-15f11100-7405-11ea-8461-ee1888fa84fd.gif)

apupier commented:
> > ```
> > * mentioning issues with other features than "create vdb files" that are tested here as I would expect it to be working as automated test integrated in the same PR and so maybe handled in this PR.
> > ```
> 
> 
> Tbh I have no clue what you mean. Could you elaborate?

The use case #1 is supposed to be only the VDB fiel creation. So I was checking manually that we covered with test only the create vdb case.
Given that tests for other commands are tested in this PR, I would expect that the functionalities are working. So I tested them and discovered several bugs with them, meaning that test coverage for them is not enough.

lhein commented:
>     * mentioning issues with other features than "create vdb files" that are tested here as I would expect it to be working as automated test integrated in the same PR and so maybe handled in this PR. Also there are some code changes in this PR for code related to them so it can also be regressions, I have not checked:
>       ** most commands are visible in command palette even though they are not working
>       ** if file is opened and modified, the deletion of a datasource entry is not working (even though a notification is mentioning that it has been deleted)
>       ** Wrong validation for data source entry:
>       *** it is forbidding to use _ (by default there is one created with name SPRING_TEIID_DATA_MONGODB_MYDATASOURCE_REMOTE_SERVER_LIST so with key REMOTE_SERVER_LIST)
>       *** it is forbidding names lower than 4 but in examples there is SSL which is 3
>       *** it is not checking for same key name with uppercase which is what needs to be checked as the value is transformed and uppercased right after. An error notification is provided when finishing the command
>       ** wrong validation for Data source creation:
>       *** it is not checking for same datasource name with uppercase which is what needs to be checked as the value is transformed and uppercased right after. An error notification is provided when finishing the command
>       ** when trying to create a datasource entry without value, there is an error. The value empty string "" should be used, like for examples.
>       ** all error logs are written on the same line making it hard to read.
>       ** when using escape when trying to modify a data source entry, there is an error notification
>       ** when there are several datasources, deleting one is deleting most of the entries of the others.

Please raise issues for your findings. This PR is about adding tests. Most of what you mention here is not about these tests but about functionality. While there might be some overlap/connect between the tests and the functionality it mostly doesn't affect this PR.

lhein commented:
> The use case #1 is supposed to be only the VDB fiel creation. So I was checking manually that we covered with test only the create vdb case.
> Given that tests for other commands are tested in this PR, I would expect that the functionalities are working. So I tested them and discovered several bugs with them, meaning that test coverage for them is not enough.

Please raise issues for your found bugs. I will then be able to extend the existing tests to also cover your findings.

lhein commented on .editorconfig:
```diff
@@ -0,0 +1,14 @@
+root = true
+
+[*]
+indent_style = tab
```
and the settings.json will automatically convert spaces to tabs?

apupier replied:
which would emphasize that there is a problem of defining the same thing in several places.

lhein commented on .editorconfig:
```diff
@@ -0,0 +1,14 @@
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.md]
+trim_trailing_whitespace = false
```
this was a default config and I didn't touch this part. We can change that if you are after it. Have no strong opinion on that.

apupier replied:
default config coming from where?

lhein replied:
From the course Brian and I mentioned several times.

apupier commented:
> > The use case #1 is supposed to be only the VDB fiel creation. So I was checking manually that we covered with test only the create vdb case.
> > Given that tests for other commands are tested in this PR, I would expect that the functionalities are working. So I tested them and discovered several bugs with them, meaning that test coverage for them is not enough.
> 
> Please raise issues for your found bugs. I will then be able to extend the existing tests to also cover your findings.

As i mentioned, there are several product code changes, not only the test is changed. They can be regressions.

Also, it means that the tests in the PR are not covering several basic cases

lhein commented:
> As i mentioned, there are several product code changes, not only the test is changed. They can be regressions.

These changes were mostly changes to make testing easier by exporting formerly internal functions and change the signature of those functions to work without the treenode objects.

lhein commented:
>     * when there are several datasources, deleting one is deleting most of the entries of the others. For instance
>       ![deleteEntriesDeltessTooManyEntries](https://user-images.githubusercontent.com/1105127/78008169-a2390080-733f-11ea-90de-c63868fae55d.gif)

Seems you discovered a regression. Please raise an issue.

apupier commented:
> > ```
> > * when there are several datasources, deleting one is deleting most of the entries of the others. For instance
> >   ![deleteEntriesDeltessTooManyEntries](https://user-images.githubusercontent.com/1105127/78008169-a2390080-733f-11ea-90de-c63868fae55d.gif)
> > ```
> 
> 
> Seems you discovered a regression. Please raise an issue.

no, this must be fixed in this PR so that the regression is not introduced in the master

lhein commented on src/commands/CreateDataSourceEntryCommand.ts:
```diff
@@ -21,12 +21,16 @@ import { IDVConfig, IDataSourceConfig, IEnv } from '../model/DataVirtModel';
+					const file: string = ds.getProject().file;
+					handleDataSourceEntryCreation(yaml, ds.dsConfig, file, eName, eValue)
 						.then( (success: boolean) => {
+							extension.dataVirtProvider.refresh();
```
out of scope of this PR. there is already an issue for that (https://github.com/jboss-fuse/vscode-datavirt/issues/27)

lhein commented on src/commands/DeleteDataSourceCommand.ts:
```diff
@@ -20,47 +20,44 @@ import * as extension from '../extension';
 			}
 		} else {
-			extension.log("handleDataSourceEdit: Unable to delete the datasource...");
+			extension.log('handleDataSourceEdit: Unable to delete the datasource...');
```
You think the user can't remember what datasource he did the delete on a few milliseconds ago?

lhein replied:
actually its the log. so it might be useful.

lhein commented on src/extension.ts:
```diff
@@ -122,16 +118,16 @@ function creatDataVirtView(): void {
 	dataVirtTreeView.onDidChangeVisibility(async () => {
 		if (dataVirtTreeView.visible === true) {
-			await dataVirtProvider.refresh().catch(err => console.log(err));
+			await dataVirtProvider.refresh().catch((err) => console.log(err));
```
paranthesis are overcomplicated? I am not going to agree here. Its a style setting I enabled for this project to always have these paranthesis. Unless we decide at some point to not do that for all of our projects I will keep it for now.

apupier replied:
ok, this is the default for ESLint https://eslint.org/docs/rules/arrow-parens

I wasn't aware of the ambiguity that can be triggered https://eslint.org/docs/rules/arrow-parens#if-statements

it would have been better to do the modification in a specific PR.

lhein commented on src/extension.ts:
```diff
@@ -141,4 +137,12 @@ function handleVisibleEditorChanges(event) {
+}
+
+export function fillDataTypes(): void {
+	DATASOURCE_TYPES.set('SpringBoot', new SpringDataSource(TEMPLATE_NAME));
+	DATASOURCE_TYPES.set('MongoDB', new MongoDBDataSource(TEMPLATE_NAME));
+	DATASOURCE_TYPES.set('Salesforce', new SalesForceDataSource(TEMPLATE_NAME));
+	DATASOURCE_TYPES.set('Google Sheets', new GoogleSheetsDataSource(TEMPLATE_NAME));
+	DATASOURCE_TYPES.set('Rest Based', new RestBasedDataSource(TEMPLATE_NAME));
```
it comes from the very first example vdb file I've been given by Ramesh as far as I can remember. Anyways this is not the scope of this PR. It has to be fixed in a separate PR for #86

lhein commented:
https://github.com/jboss-fuse/vscode-datavirt/issues/107 has been created to cover the move to ESLint

apupier commented:
> #107 has been created to cover the move to ESLint

great, we can remove all tslint stuff from this PR then

lhein commented:
that has already been done

lhein commented:
> no, this must be fixed in this PR so that the regression is not introduced in the master

fixed

apupier commented on src/test/commands.datasource.entry.test.ts:
```diff
@@ -0,0 +1,743 @@
+ */
+'use strict';
+
+import * as chai from 'chai';
+import * as sinonChai from 'sinon-chai';
+import * as vscode from 'vscode';
+import * as fs from 'fs';
+import * as path from 'path';
```
optional: better to have sorted imports

lhein replied:
I will do that once we defined a way to automate this.

apupier replied:
reported https://issues.redhat.com/browse/FUSETOOLS2-328

apupier commented on src/test/commands.datasource.test.ts:
```diff
@@ -0,0 +1,274 @@
+ */
+'use strict';
+
+import * as chai from 'chai';
+import * as sinonChai from 'sinon-chai';
+import * as vscode from 'vscode';
+import * as fs from 'fs';
+import * as path from 'path';
```
optional: better to have alphabetically sorted import

lhein replied:
I will do that once we defined a way to automate this.

apupier commented on src/test/commands.datasource.test.ts:
```diff
@@ -0,0 +1,274 @@
+const should = chai.should();
+
+describe('Commands Tests', () => {
+	let p: string;
```
what's 'p'?

apupier commented on src/test/commands.vdb.test.ts:
```diff
@@ -0,0 +1,94 @@
+const should = chai.should();
+
+describe('Commands Tests', () => {
+	let p: string;
```
what's 'p'?

apupier commented on src/test/runTest.ts:
```diff
@@ -17,6 +17,7 @@
 import * as path from 'path';
 
 import { runTests } from 'vscode-test';
+import { TestOptions } from 'vscode-test/out/runTest';
```
unused import

lhein commented on src/test/commands.datasource.entry.test.ts:
```diff
@@ -0,0 +1,743 @@
+	const mongoDSType: string = mongoTempl.type;
+	let f: string;
+	let p: string;
+	let templFolder: string;
```
I can do that but where in "temporary" will find the "L" ?

apupier replied:
when readign fast the little "l" is nto very visible. At first read, I missed it. At second read, I thought it was a typo. Then I need to jump to where it is created to understand that it was not that and finally understand that it is a templateFolder.
Using the full name remove all ambiguity.

lhein commented:
> sharing progress on review before stopping for today.
> 
>     * some of the previous feedback were not adressed/discussed. For instance https://github.com/jboss-fuse/vscode-datavirt/pull/84/files#r400715671 , https://github.com/jboss-fuse/vscode-datavirt/pull/84/files#r400715737 , https://github.com/jboss-fuse/vscode-datavirt/pull/84/files#r400716396 , https://github.com/jboss-fuse/vscode-datavirt/pull/84/files#r400718215
> 
>     * unless I missed, I have not found test for the regression that was spotted for deletion of a datasource when several datasource where in the file. It seems important to have one as there was a regression.

I addressed or commented on all of your findings and I added the regression test. Let me know when you find more issues.

apupier commented on src/commands/CreateDataSourceCommand.ts:
```diff
@@ -20,44 +20,44 @@ import * as extension from '../extension';
 			}
 		} else {
-			extension.log("handleDataSourceCreation: Unable to create the datasource because no name and type were given...");
+			extension.log('handleDataSourceCreation: Unable to create the datasource because of missing parameters...');
```
created https://issues.redhat.com/browse/FUSETOOLS2-325 as seems it won't be handled in this PR

apupier commented on src/commands/DeleteDataSourceCommand.ts:
```diff
@@ -20,47 +20,44 @@ import * as extension from '../extension';
-	handleDataSourceDeletion(ctx)
+export function deleteDataSourceCommand(ctx): void {
+	if (ctx) {
+		const dsNode: DataSourceTreeNode = ctx;
```
you provided the type in the method signature int his R. that's great. But as you made this change butnot the other one, i'm not sure I was clear enough. i was also mentioning that the line `const dsNode: DataSourceTreeNode = ctx;` which is used for casting can be removed too.

can be handled in another PR as it was already the case before.

apupier commented on src/test/commands.vdb.test.ts:
```diff
@@ -0,0 +1,100 @@
+				});
+		});
+
+		it('should not generate a VDB file when handing over invalid folder name', (done) => {
+			createVDBCommand.handleVDBCreation(undefined, name, templateFolder)
+				.then( (success) => {
+					if (success) {
+						done(new Error('Execution of the command returned true, but should not'));
+					} else {
+						done();
+					}
+				})
+				.catch( (err) => {
+					done(err);
+				});
+		});
```
what do you think to use async await instead of then?
it will allow to write:
```
		it('should return false when deleting a datasource entry with not existing key', async () => {
			let success = await deleteDSEntryCommand.handleDataSourceEntryDeletion(dvConfig, dsConfig, vdbFile, 'NOT_EXISTING');
			should.equal(false, success, 'Execution of the Edit DataSource command returned true, but it should not');
		});
```

instead of:
```
		it('should not generate a VDB file when handing over invalid folder name', (done) => {
			createVDBCommand.handleVDBCreation(undefined, name, templateFolder)
				.then( (success) => {
					if (success) {
						done(new Error('Execution of the command returned true, but should not'));
					} else {
						done();
					}
				})
				.catch( (err) => {
					done(err);
				});
		});
```
or even if factorized keeping then:
```
		it('should not generate a VDB file when handing over invalid folder name', (done) => {
			createVDBCommand.handleVDBCreation(undefined, name, templateFolder)
				.then( checkReturnFalse(done));
		});
```


after doing that it seems like we can factorize even more and provide a method checkReturnFalse(dvConfig: IDVConfig, dsConfig: IDataSourceConfig, file: string, key: string) and call only this method in several of the tests. it will reduce the test core method to a single line and focus on the provided parameters only

lhein replied:
That is more a question of what one prefers. I for myself like the Promise.then more and you seem to prefer async / await. Both are valid approaches of course.

lhein replied:
I consider your option 3 as I like it more than the 1. Will do that needed changes later on.
